capstone/src/main/java/com/google/sps/data/RecommenderAlgorithm.java

```diff
@@ -11,49 +13,158 @@
+
+        List<Key> rooms = new ArrayList<>();
+        for (Map.Entry<Key, Double> entry : distances.entrySet()) {
+            rooms = recommender.findSimilarRooms(user, entry.getKey(), rooms);
+        }
+
+        if (rooms.size() < amount) {
+            return rooms;
+        }
+
+        return rooms.subList(0, amount);
+    }
+
+    public static List<Key> findSimilarRooms(User user, Key key, List<Key> rooms) {
         DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
-        HashMap<String, Integer> userTagMap = RecommenderAlgorithm.populateUserTagMap(user);
-        
-        Query otherUserQuery = new Query(User.USER_ENTITY_NAME);
-        PreparedQuery otherUserResults = datastore.prepare(otherUserQuery);
- 
-        User otherUser = null;
-        HashMap<String, Integer> otherUserTagMap = null;
-        HashMap<User, HashMap<String,Integer>> otherUserVectorMap = new HashMap<User, HashMap<String,Integer>>();
-        for(Entity otherUserEntity : otherUserResults.asIterable()) {
-            otherUser = new User(otherUserEntity);
-            otherUserTagMap = RecommenderAlgorithm.populateUserTagMap(otherUser);
-            otherUserVectorMap.put(otherUser, otherUserTagMap);
+        Query query = new Query(Room.ROOM_ENTITY_NAME);
+        PreparedQuery results = datastore.prepare(query);
+
+        if (results.countEntities() == 0) {
+            return new ArrayList<>();
         }
- 
-        // Algorithm is work in progress. Have yet to normalize and calculate relationship.
-        return null;
+
+        for (Entity result : results.asIterable()) {
+            Room room = new Room(result);
+            if (room.getAllFollowers().contains(key) && !room.getAllFollowers().contains(user.getUserKey()) &&
+                !rooms.contains(room.getRoomKey())) {
+                rooms.add(room.getRoomKey());
+            }
+        }
+        return rooms;
     }
- 
-    public static HashMap<String, Integer> populateUserTagMap(User user) {
+
+    public static void populateEmbeddedEntity(User user) {
         DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
-        Filter userRoomFilter = new FilterPredicate(Room.FOLLOWERS_PROPERTY_KEY, FilterOperator.EQUAL, user.getUserKey());
-        Query userRoomQuery = new Query(Room.ROOM_ENTITY_NAME).setFilter(userRoomFilter);
-        PreparedQuery userRoomResults = datastore.prepare(userRoomQuery);
- 
-        HashMap<String, Integer> userTagMap = Tag.constructUserVectorMap();
-        List<String> userTags = null;
-        for(Entity roomEntity : userRoomResults.asIterable()){
-            userTags = new Room(roomEntity).getAllTags();
-            for(String tag : userTags) {
-                userTagMap.put(tag, userTagMap.get(tag) + 1);
+        Query query = new Query(Room.ROOM_ENTITY_NAME);
+        PreparedQuery results = datastore.prepare(query);
+        EmbeddedEntity entity = user.getEmbeddedTags();
```

Comment:
Why was the implementation which used a filter removed? Querying all rooms will not scale, and a filter will use indexes to make getting the rooms you are looking for quick.

Reply:
gotcha... I didn't even notice that Andrew had gotten the filter to work. Didn't think that filter would do anything because I thought it would check if the rooms followers list was equal to the user but I guess it checks if the user is in the rooms followers list